### PR TITLE
authentication - verify and decode jwt

### PR DESCRIPTION
**Pull request by andreakreichgauer**
This PR introduces authentication interceptor for incoming requests.
 
All incoming requests need to have a `Authorization` Header with a valid Bearer token in order to be processed or otherwise will be refused with HTTP 401 as a response.

For the moment a `"X-Forwarded-User"` Header works as well to pass authorization, but will be removed at some point.

### 🧰 Technical Solution
- [x] add authentication interceptor to check `Authorization` Header or if none present `X-Forwarded-User` Header instead
- [x] verify Bearer token from Authorization Header
- [x] decode Bearer token to jwt
- [x] add `authentication_interceptor.enabled` (default=true) and `sso_realm.address` to hapi.properties

### ‼ Known Issues
To test this feature locally `authentication_interceptor.enabled must be set to true (default) and `sso_realm.address=` must be provided in hapi.properties or as env var `HAPI_SSO_REALM_ADDRESS`

You can use [this postman collection](https://dot-base.postman.co/workspace/4909cdbe-489d-41fc-8ea9-8326c7c79fda/pull-request/a045d794-c084-44e4-bb90-b1ea14a1b57f?workspace=4909cdbe-489d-41fc-8ea9-8326c7c79fda) and [this realm address](https://sso.movebase.charite.de/auth/realms/dotbase) to test the feature locally.

### CODE REVIEW

**Review by friedow — APPROVED** [review]
How do I get the public key of a keycloak realm?

File: src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java [hunk 1]
```diff
@@ -313,6 +315,10 @@ public static String getServerAddress() {
     return HapiProperties.getProperty(SERVER_ADDRESS);
   }
 
+  public static String getRealmPublicKey() {
```
Comment (friedow) [inline src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java:318]: Is there any way we can determine the public key of a realm before starting the keycloak server itself? If not, we've got a deployment problem here. 

To set `IDENTITY_PROVIDER_REALM_PUBLIC_KEY` we need to have a running keycloak instance. The problem is, all services start at the same time. We would have to manually get the realm public key from `<your-sso-url>/auth/realms/<realm-name>` and set it on the fhir server as an env variable after the service has already started. This forces the service to restart.

I see two options to resolve this problem:
1. Retrieve the public key automatically from the keycloak server. I'd set an environment variable like `KEYCLOAK_URL`.
2. Use a key that can be determined before the keycloak server starts. This way we could set this on the inital stack launch.

File: src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java [hunk 2]
```diff
@@ -600,6 +606,10 @@ public static boolean isFhirPathFilterInterceptorEnabled() {
     return HapiProperties.getBooleanProperty("fhirpath_interceptor.enabled", false);
   }
 
+  public static boolean isAuthenticationInterceptorEnabled() {
+    return HapiProperties.getBooleanProperty(AUTHENTICATION_ENABLED, false);
```
Comment (friedow) [inline src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java:610]: Since this a security feature, I think it should be enabled by default.

**Review by friedow — CHANGES_REQUESTED** [review]

File: src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java (hunk 1 above)
Comment (andreakreichgauer) [reply src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java]: Yes, that makes sense for the deployment. I changed it to option 1. The Public Key is retrieved from Keycloak once the fhir server comes up

File: src/main/java/ca/uhn/fhir/jpa/starter/dotBase/api/IdentityProvider.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      HttpRequest request = HttpRequest
+        .newBuilder()
+        .uri(URI.create(realm))
+        .setHeader("Accept", "application/json")
+        .GET()
+        .build();
+      HttpResponse<String> response = HTTP_CLIENT.send(request, BodyHandlers.ofString());
+      return (response.statusCode() == 200 ? publicKey(response.body()) : null);
+    } catch (IOException | InterruptedException | JSONException e) {
+      ourLog.info(e.getMessage());
```
Comment (friedow) [inline src/main/java/ca/uhn/fhir/jpa/starter/dotBase/api/IdentityProvider.java:41]: Is this intentionally an `.info` call or did you mean to call `.error` here?

**Review by friedow — DISMISSED** [review]

**Review by friedow — APPROVED** [review]

File: src/main/java/ca/uhn/fhir/jpa/starter/dotBase/api/IdentityProvider.java (hunk 3 above)
Comment (andreakreichgauer) [reply src/main/java/ca/uhn/fhir/jpa/starter/dotBase/api/IdentityProvider.java]: good point. This should be ourLog.error

**Pull request merged**